### PR TITLE
Drop support for DBAL 3

### PR DESCRIPTION
**Pull request by derrabus**
This PR removes all compatibility code for DBAL 3. ~~I'm flagging it as a draft for now because I believe that we can backport a few things.~~

### CODE REVIEW

**Comment by greg0ire** [comment]
There is code here that could be simplified I think: https://github.com/doctrine/orm/blob/3907872046b9a697860f51266714c2eb8e3b19d0/src/Mapping/ClassMetadataFactory.php#L627-L660

File: src/Query.php [hunk 1]
```diff
@@ -635,9 +633,9 @@ public function setLockMode(LockMode|int $lockMode): self
     /**
      * Get the current lock mode for this query.
      *
-     * @return int|null The current lock mode of this query or NULL if no specific lock mode is set.
+     * @return LockMode|null The current lock mode of this query or NULL if no specific lock mode is set.
      */
-    public function getLockMode(): int|null
+    public function getLockMode(): LockMode|null
```
Comment (derrabus) [inline src/Query.php]: Bugfix backported as #11462.

File: src/UnitOfWork.php [hunk 2]
```diff
@@ -2653,7 +2640,7 @@ private function eagerLoadCollections(array $collections, ToManyInverseSideMappi
                 $entities[] = $collection->getOwner();
             }
 
-            $found = $this->getEntityPersister($targetEntity)->loadAll([$mappedBy => $entities], $mapping['orderBy'] ?? null);
+            $found = $this->getEntityPersister($targetEntity)->loadAll([$mappedBy => $entities], $mapping->orderBy());
```
Comment (derrabus) [inline src/UnitOfWork.php]: Backported to #11464

File: tests/Tests/ORM/Cache/Persister/Entity/EntityPersisterTestCase.php [hunk 3]
```diff
@@ -97,7 +98,7 @@ public function testInvokeGetSelectSQL(): void
             ->with(
                 self::identicalTo(['name' => 'Foo']),
                 self::identicalTo($associationMapping),
-                self::identicalTo(1),
+                self::identicalTo(LockMode::OPTIMISTIC),
```
Comment (derrabus) [inline tests/Tests/ORM/Cache/Persister/Entity/EntityPersisterTestCase.php]: Backported as #11465.

**Comment by derrabus** [comment]
PR is ready.

**Comment by greg0ire** [comment]
Doing a case insensitive search for `dbal 3`, this comes up: https://github.com/doctrine/orm/blob/dbfe47b07bb1ee858735a50f46736d336164be1d/tests/Tests/ORM/Mapping/ClassMetadataFactoryTest.php#L128-L142

**Comment by derrabus** [comment]
I must've missed that one. I'll remove the test.

**Review by greg0ire — APPROVED** [review]

**Pull request merged**